### DISCUSSION TITLE
Sorted Tables in Metadata (#~ stream)

### DISCUSSION THREAD

**Discussion by AlexanderMorou** [dotnet/roslyn#60576] [Q&A] [answered]
### `#~` stream ###
I'm trying to update an older project to .NET 6 that handled reading .NET Metadata on its own.

In trying to understand the data, I came across the `Sorted` field of the `#~` section of metadata. Sorted is presented in the `ECMA-335 spec, 6th edition, June 2012`, it is in section `II.24.2.6`, page 273.

Using file `System.Private.CoreLib.dll` from `dotnet-sdk-5.0.100-win-x64`, I noticed that flags 50, 54 and 55 were set, but there doesn't appear to be a description for those tables.

While I could do a deep dive into the roslyn code base, that doesn't feel like a productive use of time. I think that because I don't know the first thing about the code base. I'd have to download it all, and I honestly don't even know where to begin looking.

### The simplest form of my question boils down to: ###
Is there a clear-cut description of the ECMA-335 for modern .NET 6, or has the metadata definition not changed?

#### Context on use case: ####
Being able to load multiple .NET Runtimes to run comparisons between the various libraries. Such as what version a particular type was introduced. Then storing the computed results as an XML document. 

While I know this is all contained in the .NET Documentation, having that awareness at runtime is often useful in certain code gen scenarios.

**Comment by PathogenDavid** [comment, marked as answer]
This might be a better question for [the runtime repo](https://github.com/dotnet/runtime).

.NET has evolved ECMA-335 over the years. Augmentations to it are [documented over in the runtime repo](https://github.com/dotnet/runtime/blob/4fed599fe7b7ebea8886d3c987c45e7520ba0fa3/docs/design/specs/Ecma-335-Augments.md). (I'm not entirely sure how complete that document is or if it has the info you're looking for.)

> Being able to load multiple .NET Runtimes to run comparisons between the various libraries. Such as what version a particular type was introduced. Then storing the computed results as an XML document.

You might look into moving towards using [`System.Reflection.Metadata.Ecma335`](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.metadata.ecma335) instead of reading things manually. (I believe it's what Roslyn uses, so at the very least [its source code](https://github.com/dotnet/runtime/blob/7281406fe89936e3bedd5e3a1c742f421f21a9f3/src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/MetadataBuilder.cs) is probably more relevant to you than anything in this repo.)

You may find [the API diffs published with every release](https://github.com/dotnet/core/tree/main/release-notes/6.0/api-diff) to be helpful. I believe these are generated at least in part via [Microsoft.DotNet.AsmDiff](https://github.com/dotnet/arcade/tree/main/src/Microsoft.DotNet.AsmDiff).

**Reply by AlexanderMorou** [reply]
~~Seems I'll need to repost in [the runtime repo](https://github.com/dotnet/runtime).~~

Actually you answered my question by [this link](https://github.com/dotnet/runtime/blob/4fed599fe7b7ebea8886d3c987c45e7520ba0fa3/docs/design/specs/Ecma-335-Augments.md).
